File: packages/ai/sandbox-assets/fretik_apps/objects.py
# Hand-maintained (NOT manifest-generated). Copied verbatim into
# fretik_apps/ by scripts/generate-sdk.ts. Edit here, then `bun run gen:sdk`.

"""fretik_apps.objects — code-mode SDK for the team's ontology (object types +
records). The bulk / migration power path.

Use it when a task touches MANY records or restructures a type — insert hundreds
of rows, move records between types, merge/split types, data-preserving retype.
Everything runs SERVER-SIDE in fretik-backend: field validation, team scoping,
grants and the domain-events journal are applied exactly like the manageRecord /
manageObjectType tools. Intermediate data you build here (parsed files, mappings)
never re-enters your context — only the small result summary does.

For ONE record or a single interactive edit, use the manageRecord /
manageObjectType / manageField tools instead; this SDK is the batch path.

  from fretik_apps import objects

  # Bulk insert — ids[i] aligns with rows[i] (None if that row failed).
  res = objects.records.bulk_create("clients", [
      {"name": "ACME", "vat": "FR123"},
      {"name": "Globex"},
  ])
  print(res["okCount"], "created,", len(res["errors"]), "failed")

  # Migration sketch: split one type into two
  objects.schema.create_type("supplier", "Supplier", "Companies we buy from", fields=[
      {"label": "Name", "type": "text", "is_title": True, "description": "Supplier name"},
  ])
  page = objects.records.query("contact", filters={"kind": "supplier"})
  created = objects.records.bulk_create(
      "supplier", [{"name": r["data"]["name"]} for r in page["records"]]
  )
  objects.records.bulk_delete([r["id"] for r in page["records"]])
"""

from typing import Any

from ._runtime import _call_objects

# Field dicts use Python snake_case; the backend wants camelCase. Map only the
# multi-word keys — single-word ones (label, type, description, config) pass
# through untouched.
_FIELD_KEY_MAP = {
    "is_title": "isTitle",
    "display_in_filters": "displayInFilters",
}

# Relation dicts use snake_case; the backend wants camelCase.
_RELATION_KEY_MAP = {
    "relation_key": "relationKey",
    "link_type_id": "linkTypeId",
    "to_record_id": "toRecordId",
    "to_document_id": "toDocumentId",
}


def _clean(args: dict[str, Any]) -> dict[str, Any]:
    """Drop None values so optional args fall back to their server defaults."""
    return {k: v for k, v in args.items() if v is not None}


def _field(spec: dict[str, Any]) -> dict[str, Any]:
    """Normalize one field spec (snake_case → the backend's camelCase keys)."""
    return {_FIELD_KEY_MAP.get(k, k): v for k, v in spec.items() if v is not None}


def _relation(spec: dict[str, Any]) -> dict[str, Any]:
    """Normalize one relation spec (snake_case → camelCase)."""
    return {
        _RELATION_KEY_MAP.get(k, k): v for k, v in spec.items() if v is not None
    }


def _row(row: dict[str, Any]) -> dict[str, Any]:
    """Normalize a bulk_create row to {"data", "relations"}. A bare field map IS
    the record's data; pass {"data": {...}, "relations": [...]} to attach
    outgoing relations to the new record in the same write."""
    if "data" in row and isinstance(row["data"], dict):
        data, rels = row["data"], row.get("relations") or []
    else:
        data, rels = row, []
    out: dict[str, Any] = {"data": data}
    if rels:
        out["relations"] = [_relation(r) for r in rels]
    return out


class _Records:
    """Bulk record operations. Each call is ONE backend round-trip that fans out
    to a set-based write — never a row-by-row loop."""

    def bulk_create(
        self, type_key: str, rows: list[dict[str, Any]]
    ) -> dict[str, Any]:
        """Create many records of `type_key`. Each row is a field map
        (key → value), validated against the type's schema server-side.

        To attach outgoing relations in the same write, give a row as
        {"data": {<field map>}, "relations": [{"relation_key": "client",
        "to_record_id": "<id>"}]} — target by `to_record_id` or an uploaded
        file's `to_document_id` (its document record).

        Returns {"ids": [...], "okCount": int, "errors": [{index, error}],
        "relationErrors": [{index, error}]}. `ids[i]` is the new id for `rows[i]`
        (None if it failed); `relationErrors` is indexed by row. Keep the result
        in a variable and print only counts, not the whole list.
        """
        return _call_objects(
            "records.bulk_create",
            {"typeKey": type_key, "rows": [_row(r) for r in rows]},
        )

    def bulk_update(
        self,
        updates: list[dict[str, Any]] | None = None,
        *,
        merge: bool = True,
        records: list[dict[str, Any]] | None = None,
        type_key: str | None = None,
    ) -> dict[str, Any]:
        """Update the data of many records. Each item is
        {"id": "<record id>", "data": {<field map>}}. No type_key needed —
        each id routes itself. Records outside your team are skipped.

        merge=True (default): PATCH — only the keys you pass change, the rest
        are kept; pass a key with value None to clear it. merge=False: full
        replace — omitted keys are cleared.

        Returns {"updatedIds": [...], "okCount": int, "errors": [{id, error}]}.
        """
        # Tolerant of the bulk_create call shape: `records=` aliases `updates`,
        # and a stray `type_key` is accepted (each id routes itself).
        items = updates if updates is not None else records
        if items is None:
            raise TypeError(
                "bulk_update expects a list of {'id', 'data'} updates"
            )
        return _call_objects(
            "records.bulk_update", {"updates": items, "merge": merge}
        )

    def bulk_delete(self, record_ids: list[str]) -> dict[str, Any]:
        """Delete many records by id. Ids outside your team are skipped.

        Returns {"deletedIds": [...], "okCount": int, "errors": [{id, error}]}.
        """
        return _call_objects(
            "records.bulk_delete", {"recordIds": record_ids}
        )

    def query(
        self,
        type_key: str,
        filters: dict[str, Any] | None = None,
        page: int = 0,
        limit: int = 200,
    ) -> dict[str, Any]:
        """Read a page of the team's confirmed records of `type_key`, each with
        its `data` map. `filters` is an equality map on field keys. The
        migration read primitive — fetch a batch, transform in-sandbox, write it
        back. Full-text search stays in the SQL tool.

        Returns {"records": [{"id", "label", "status", "data"}]}.
        """
        return _call_objects(
            "records.query",
            _clean(
                {
                    "typeKey": type_key,
                    "filters": filters,
                    "page": page,
                    "limit": limit,
                }
            ),
        )


class _Schema:
    """Object-type & field migrations. Field changes are DDL — bounded in count,
    so these are normal per-call ops (not thousands-scale like records)."""

    def create_type(
        self,
        key: str,
        label: str,
        description: str,
        label_plural: str | None = None,
        icon: str | None = None,
        sharing: dict[str, Any] | None = None,
        fields: list[dict[str, Any]] | None = None,
    ) -> dict[str, Any]:
        """Create an object type, provisioning its typed table. `description` is
        one line — what the type is for (required; you read it later as ground
        truth). Pass `fields` to build the whole schema in ONE call (each a dict
        like {"label": "Name", "type": "text", "is_title": True, "description":
        "..."}; every field needs its own one-line `description`). Max 30 fields
        per type. Exclude relation/rollup fields — add those with `add_field`.
        Colors are auto-assigned; a select/multi_select option may set an
        optional `color` (a palette token) to override.

        `sharing` sets the cross-team audience (owner team only). Omit = internal
        (owning team only). {"mode": "org", "permission": "read"} shares with the
        whole organization; {"mode": "teams", "teams": [{"teamId": "...",
        "permission": "read"}]} with specific teams. Records inherit this live.

        Returns {"id", "key", "fields": [{key, type}]}.
        """
        if fields and len(fields) > 30:
            raise ValueError(
                f"create_type: max 30 fields per type, got {len(fields)}"
            )
        return _call_objects(
            "schema.create_type",
            _clean(
                {
                    "key": key,
                    "label": label,
                    "labelPlural": label_plural,
                    "description": description,
                    "icon": icon,
                    "sharing": sharing,
                    "fields": [_field(f) for f in fields] if fields else None,
                }
            ),
        )

    def update_type(
        self,
        type_key: str,
        label: str | None = None,
        label_plural: str | None = None,
        description: str | None = None,
        icon: str | None = None,
        enabled: bool | None = None,
        sharing: dict[str, Any] | None = None,
        add_fields: list[dict[str, Any]] | None = None,
    ) -> dict[str, Any]:
        """Patch a type's metadata AND/OR add several new fields in one call
        (`add_fields`, same shape as `create_type`'s `fields` — each needs a
        one-line `description`). Editing or removing existing fields is
        `change_field`. `sharing` changes the cross-team audience (owner team
        only; same shape as `create_type`).

        Returns {"key", "addedFields": [{key, type}]}.
        """
        return _call_objects(
            "schema.update_type",
            _clean(
                {
                    "typeKey": type_key,
                    "label": label,
                    "labelPlural": label_plural,
                    "description": description,
                    "icon": icon,
                    "enabled": enabled,
                    "sharing": sharing,
                    "addFields": (
                        [_field(f) for f in add_fields] if add_fields else None
                    ),
                }
            ),
        )

    def add_field(
        self,
        type_key: str,
        label: str,
        type: str,
        description: str,
        config: dict[str, Any] | None = None,
        display_in_filters: bool | None = None,
    ) -> dict[str, Any]:
        """Add one field (ALTER TABLE ADD COLUMN) to an existing type.
        `description` is one line — what the field holds (required).

        Returns {"key", "type"}.
        """
        return _call_objects(
            "schema.add_field",
            _clean(
                {
                    "typeKey": type_key,
                    "label": label,
                    "type": type,
                    "description": description,
                    "config": config,
                    "displayInFilters": display_in_filters,
                }
            ),
        )

    def change_field(
        self,
        type_key: str,
        field_key: str,
        action: str,
        label: str | None = None,
        description: str | None = None,
        config: dict[str, Any] | None = None,
        type: str | None = None,
        display_in_filters: bool | None = None,
        enabled: bool | None = None,
        cascade: bool | None = None,
    ) -> dict[str, Any]:
        """Edit one field. `action` is "update" (keeps values), "changeType"
        (pass `type`; RESETS the field's values), or "delete" (pass cascade=True
        to drop a field that holds values).
        """
        return _call_objects(
            "schema.change_field",
            _clean(
                {
                    "typeKey": type_key,
                    "fieldKey": field_key,
                    "action": action,
                    "label": label,
                    "description": description,
                    "config": config,
                    "type": type,
                    "displayInFilters": display_in_filters,
                    "enabled": enabled,
                    "cascade": cascade,
                }
            ),
        )

    def delete_type(self, type_key: str) -> dict[str, Any]:
        """Drop a type and every record in it — the last step of a merge/split."""
        return _call_objects("schema.delete_type", {"typeKey": type_key})


records = _Records()
schema = _Schema()

File: packages/shared/src/services/external-apps/exec/objects.ts
import { z } from "zod";
import { MAX_BULK_ITEMS } from "../../../lib/db-bulk";
import {
  fieldConfigSchema,
  fieldDefinitionTypeSchema,
} from "../../../schemas/field-definitions";
import { audienceSchema } from "../../../schemas/object-sharing";
import { recordRelationInputSchema } from "../../../schemas/ontology";
import type { EventActor } from "../../domain-events/emit";
import { FIELD_DEFINITION_LIMITS } from "../../field-definitions/constants";
import { createFieldDefinition } from "../../field-definitions/create";
import { deleteFieldDefinition } from "../../field-definitions/delete";
import { getFieldDefinitionsForTeam } from "../../field-definitions/get-for-team";
import { updateFieldDefinition } from "../../field-definitions/update";
import { bulkCreateObjectRecords } from "../../object-records/bulk-create";
import { bulkDeleteObjectRecords } from "../../object-records/bulk-delete";
import { bulkUpdateObjectRecords } from "../../object-records/bulk-update";
import { queryObjectRecords } from "../../object-records/query";
import { OBJECT_TYPE_LIMITS } from "../../object-types/constants";
import { createObjectType } from "../../object-types/create";
import { createObjectTypeWithFields } from "../../object-types/create-with-fields";
import { deleteObjectType } from "../../object-types/delete";
import { resolveObjectTypeId } from "../../object-types/resolve";
import { updateObjectType } from "../../object-types/update";
import type { ExecContext, SandboxExecResponse } from "./types";

/**
 * `kind: "objects"` dispatch for `POST /sandbox/exec` — the server side of the
 * Python `fretik_apps.objects` SDK (code-mode). The bulk / migration power path:
 * the agent writes ONE in-sandbox script (create a type, move records between
 * types, insert hundreds of rows) and only a small result summary re-enters its
 * context.
 *
 * Every op goes through the SAME validated shared services as the domain tools
 * (`manageRecord` / `manageObjectType` / `manageField`), so field validation,
 * the typed extension table, and `domain_events` stay consistent and auditable.
 * Record ops use the batch services (`bulk{Create,Update,Delete}ObjectRecords`)
 * — one set-based statement per chunk, no per-row SQL. Tenancy is pinned by the
 * sandbox JWT: type resolution is team-scoped and the bulk services drop any
 * record id not owned by the JWT's team, so a script can never reach another
 * team's data.
 */
export const dispatchObjects = async (
  ctx: ExecContext,
  op: string,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const actor: EventActor = {
    actorType: "agent",
    actorUserId: ctx.userId,
    conversationId: ctx.conversationId,
  };

  try {
    switch (op) {
      case "records.bulk_create":
        return await bulkCreate(ctx, actor, rawArgs);
      case "records.bulk_update":
        return await bulkUpdate(ctx, actor, rawArgs);
      case "records.bulk_delete":
        return await bulkDelete(ctx, actor, rawArgs);
      case "records.query":
        return await queryRecords(ctx, rawArgs);
      case "schema.create_type":
        return await createType(ctx, rawArgs);
      case "schema.update_type":
        return await updateType(ctx, rawArgs);
      case "schema.add_field":
        return await addField(ctx, rawArgs);
      case "schema.change_field":
        return await changeField(ctx, rawArgs);
      case "schema.delete_type":
        return await deleteType(ctx, rawArgs);
      default:
        return { status: "error", message: `Unknown objects op: ${op}` };
    }
  } catch (error) {
    return { status: "error", message: errMsg(error) };
  }
};

// ── Record ops (fully batched — one set-based statement per chunk) ─────

const bulkCreateArgs = z.object({
  typeKey: z.string().min(1).max(60),
  // Each row is its field `data` plus optional outgoing `relations`. Created in
  // one batched pass (records, then their links).
  rows: z
    .array(
      z.object({
        data: z.record(z.string(), z.unknown()),
        relations: z.array(recordRelationInputSchema).optional(),
      }),
    )
    .min(1)
    .max(MAX_BULK_ITEMS),
});

const bulkCreate = async (
  ctx: ExecContext,
  actor: EventActor,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const { typeKey, rows } = bulkCreateArgs.parse(rawArgs);
  const objectTypeId = await resolveTeamType(ctx, typeKey);
  if (objectTypeId === null) return unknownType(typeKey);

  const result = await bulkCreateObjectRecords({
    organizationId: ctx.organizationId,
    teamId: ctx.teamId,
    userId: ctx.userId,
    objectTypeId,
    rows,
    actor,
  });
  return {
    status: "ok",
    data: {
      ids: result.ids,
      okCount: result.ids.filter((id) => id !== null).length,
      errors: result.errors,
      relationErrors: result.relationErrors,
    },
  };
};

const bulkUpdateArgs = z.object({
  updates: z
    .array(
      z.object({
        id: z.string().min(1),
        data: z.record(z.string(), z.unknown()),
      }),
    )
    .min(1)
    .max(MAX_BULK_ITEMS),
  merge: z.boolean().optional(),
});

const bulkUpdate = async (
  ctx: ExecContext,
  actor: EventActor,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const { updates, merge } = bulkUpdateArgs.parse(rawArgs);
  const result = await bulkUpdateObjectRecords({
    teamId: ctx.teamId,
    updates,
    merge,
    actor,
  });
  return {
    status: "ok",
    data: {
      updatedIds: result.updatedIds,
      okCount: result.updatedIds.length,
      errors: result.errors,
    },
  };
};

const bulkDeleteArgs = z.object({
  recordIds: z.array(z.string().min(1)).min(1).max(MAX_BULK_ITEMS),
});

const bulkDelete = async (
  ctx: ExecContext,
  actor: EventActor,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const { recordIds } = bulkDeleteArgs.parse(rawArgs);
  const result = await bulkDeleteObjectRecords({
    teamId: ctx.teamId,
    ids: recordIds,
    actor,
  });
  return {
    status: "ok",
    data: {
      deletedIds: result.deletedIds,
      okCount: result.deletedIds.length,
      errors: result.errors,
    },
  };
};

const queryArgs = z.object({
  typeKey: z.string().min(1).max(60),
  filters: z.record(z.string(), z.unknown()).optional(),
  page: z.number().int().min(0).optional(),
  limit: z.number().int().min(1).max(500).optional(),
});

const queryRecords = async (
  ctx: ExecContext,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const { typeKey, filters, page, limit } = queryArgs.parse(rawArgs);
  const objectTypeId = await resolveTeamType(ctx, typeKey);
  if (objectTypeId === null) return unknownType(typeKey);

  const records = await queryObjectRecords({
    teamId: ctx.teamId,
    objectTypeId,
    filters,
    page,
    limit: limit ?? 200,
  });
  return {
    status: "ok",
    data: {
      records: records.map((r) => ({
        id: r.id,
        label: r.label,
        status: r.status,
        data: r.data,
      })),
    },
  };
};

// ── Schema ops ────────────────────────────────────────────────────────
//
// Field changes are DDL (`ALTER TABLE`) — inherently one statement per column,
// so a multi-field op runs them in sequence. That is NOT the "SQL in a loop"
// hazard the bulk record path avoids: field counts are tiny (dozens at most),
// and the win here is one SDK round-trip, not one DB statement.

const fieldInputSchema = z.object({
  label: z.string().min(1),
  type: fieldDefinitionTypeSchema,
  // One line — what this field holds. Required (the agent reads it as ground truth).
  description: z
    .string()
    .min(1)
    .max(FIELD_DEFINITION_LIMITS.MAX_DESCRIPTION_CHARS),
  config: fieldConfigSchema.optional(),
  isTitle: z.boolean().optional(),
});

const createTypeArgs = z.object({
  key: z.string().min(1).max(60),
  label: z.string().min(1),
  labelPlural: z.string().nullish(),
  // One line — what this type is for. Required on create.
  description: z.string().min(1).max(OBJECT_TYPE_LIMITS.MAX_DESCRIPTION_CHARS),
  icon: z.string().nullish(),
  // Cross-team audience. Omit = internal (owning team only).
  sharing: audienceSchema.optional(),
  fields: z.array(fieldInputSchema).optional(),
});

/**
 * Create an object type (provisions its typed table). Pass `fields` to build the
 * whole schema atomically — the migration entry point when splitting one type
 * into several. Colors are auto-assigned server-side.
 */
const createType = async (
  ctx: ExecContext,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const args = createTypeArgs.parse(rawArgs);
  if (args.fields && args.fields.length > 0) {
    const created = await createObjectTypeWithFields({
      organizationId: ctx.organizationId,
      teamId: ctx.teamId,
      key: args.key,
      label: args.label,
      labelPlural: args.labelPlural ?? null,
      description: args.description ?? null,
      icon: args.icon ?? null,
      color: null,
      sharing: args.sharing,
      createdByUserId: ctx.userId,
      fields: args.fields.map((f) => ({
        label: f.label,
        type: f.type,
        description: f.description ?? null,
        config: f.config,
        isTitle: f.isTitle,
      })),
    });
    return {
      status: "ok",
      data: {
        id: created.id,
        key: created.key,
        fields: created.fieldDefinitions.map((f) => ({
          key: f.key,
          type: f.type,
        })),
      },
    };
  }
  const type = await createObjectType({
    organizationId: ctx.organizationId,
    teamId: ctx.teamId,
    key: args.key,
    label: args.label,
    labelPlural: args.labelPlural ?? null,
    description: args.description ?? null,
    icon: args.icon ?? null,
    color: null,
    sharing: args.sharing,
    createdByUserId: ctx.userId,
  });
  return { status: "ok", data: { id: type.id, key: type.key, fields: [] } };
};

const updateTypeArgs = z.object({
  typeKey: z.string().min(1).max(60),
  label: z.string().optional(),
  labelPlural: z.string().nullish(),
  description: z
    .string()
    .max(OBJECT_TYPE_LIMITS.MAX_DESCRIPTION_CHARS)
    .nullish(),
  icon: z.string().nullish(),
  enabled: z.boolean().optional(),
  // Change the cross-team audience (owner team only).
  sharing: audienceSchema.optional(),
  addFields: z.array(fieldInputSchema).optional(),
});

/**
 * Update an object type in one call: patch its presentation/lifecycle metadata
 * AND/OR add several new fields at once (`addFields`). Editing or removing
 * existing fields is `change_field`; this op only grows the schema.
 */
const updateType = async (
  ctx: ExecContext,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const args = updateTypeArgs.parse(rawArgs);
  const objectTypeId = await resolveTeamType(ctx, args.typeKey);
  if (objectTypeId === null) return unknownType(args.typeKey);

  const hasMetadata =
    args.label !== undefined ||
    args.labelPlural !== undefined ||
    args.description !== undefined ||
    args.icon !== undefined ||
    args.enabled !== undefined;
  if (hasMetadata || args.sharing) {
    await updateObjectType({
      id: objectTypeId,
      patch: {
        label: args.label,
        labelPlural: args.labelPlural,
        description: args.description,
        icon: args.icon,
        enabled: args.enabled,
      },
      sharing: args.sharing,
      callerTeamId: ctx.teamId,
      createdByUserId: ctx.userId,
    });
  }

  const addedFields: { key: string; type: string }[] = [];
  for (const f of args.addFields ?? []) {
    const field = await createFieldDefinition({
      organizationId: ctx.organizationId,
      teamId: ctx.teamId,
      objectTypeId,
      label: f.label,
      type: f.type,
      config: f.config,
      description: f.description ?? null,
      isTitle: f.isTitle,
    });
    addedFields.push({ key: field.key, type: field.type });
  }

  return { status: "ok", data: { key: args.typeKey, addedFields } };
};

const addFieldArgs = z.object({
  typeKey: z.string().min(1).max(60),
  label: z.string().min(1),
  type: fieldDefinitionTypeSchema,
  // One line — what this field holds. Required.
  description: z
    .string()
    .min(1)
    .max(FIELD_DEFINITION_LIMITS.MAX_DESCRIPTION_CHARS),
  config: fieldConfigSchema.optional(),
});

const addField = async (
  ctx: ExecContext,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const args = addFieldArgs.parse(rawArgs);
  const objectTypeId = await resolveTeamType(ctx, args.typeKey);
  if (objectTypeId === null) return unknownType(args.typeKey);

  const field = await createFieldDefinition({
    organizationId: ctx.organizationId,
    teamId: ctx.teamId,
    objectTypeId,
    label: args.label,
    type: args.type,
    config: args.config,
    description: args.description ?? null,
  });
  return { status: "ok", data: { key: field.key, type: field.type } };
};

const changeFieldArgs = z.object({
  typeKey: z.string().min(1).max(60),
  fieldKey: z.string().min(1).max(60),
  action: z.enum(["update", "changeType", "delete"]),
  label: z.string().nullish(),
  description: z
    .string()
    .max(FIELD_DEFINITION_LIMITS.MAX_DESCRIPTION_CHARS)
    .nullish(),
  config: fieldConfigSchema.optional(),
  type: fieldDefinitionTypeSchema.optional(),
  enabled: z.boolean().optional(),
  cascade: z.boolean().optional(),
});

/**
 * Edit one field. `update` keeps stored values; `changeType` RESETS them
 * (add-new-column → drop-old, no in-place ALTER TYPE); `delete` drops the column
 * (`cascade` required when it holds values).
 */
const changeField = async (
  ctx: ExecContext,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const args = changeFieldArgs.parse(rawArgs);
  const objectTypeId = await resolveTeamType(ctx, args.typeKey);
  if (objectTypeId === null) return unknownType(args.typeKey);

  const fields = await getFieldDefinitionsForTeam({
    teamId: ctx.teamId,
    objectTypeId,
    includeDisabled: true,
  });
  const field = fields.find((f) => f.key === args.fieldKey);
  if (!field) {
    return {
      status: "error",
      message: `No field '${args.fieldKey}' on type '${args.typeKey}'.`,
    };
  }

  if (args.action === "delete") {
    const result = await deleteFieldDefinition({
      id: field.id,
      cascade: args.cascade ?? false,
    });
    return { status: "ok", data: result };
  }

  if (args.action === "changeType") {
    if (!args.type) {
      return { status: "error", message: "changeType requires type." };
    }
    const updated = await updateFieldDefinition({
      id: field.id,
      cascade: true,
      patch: { type: args.type, config: args.config },
    });
    return { status: "ok", data: { key: updated.key, type: updated.type } };
  }

  const updated = await updateFieldDefinition({
    id: field.id,
    patch: {
      label: args.label ?? undefined,
      description: args.description,
      config: args.config,
      enabled: args.enabled,
    },
  });
  return { status: "ok", data: { key: updated.key, type: updated.type } };
};

const deleteTypeArgs = z.object({ typeKey: z.string().min(1).max(60) });

/** Drop a type and every record in it. The last step of a merge/split. */
const deleteType = async (
  ctx: ExecContext,
  rawArgs: Record<string, unknown>,
): Promise<SandboxExecResponse> => {
  const { typeKey } = deleteTypeArgs.parse(rawArgs);
  const objectTypeId = await resolveTeamType(ctx, typeKey);
  if (objectTypeId === null) return unknownType(typeKey);

  const result = await deleteObjectType({ id: objectTypeId });
  return { status: "ok", data: result };
};

// ── Helpers ───────────────────────────────────────────────────────────

const resolveTeamType = (
  ctx: ExecContext,
  key: string,
): Promise<string | null> =>
  resolveObjectTypeId({
    organizationId: ctx.organizationId,
    teamId: ctx.teamId,
    key,
  });

const unknownType = (key: string): SandboxExecResponse => ({
  status: "error",
  message: `No object type '${key}' for this team. List types first.`,
});

const errMsg = (error: unknown): string =>
  error instanceof Error ? error.message : String(error);
